<a href="https://colab.research.google.com/github/dipayan10/Recommender-Systems/blob/main/Recommender_systems.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Example of a simple Recommendation Engine

the pourpose of this notebook is to demomstrate different types of recommendation engines. In this notebook we will recommend movies to users based on ratings, personal choices, popularity among other users etc.

####There are 3 types of recommendation engines:

1) Simple recommendation Engine: simple recommenders are basic systems that recommend the top items based on a certain metric or score.

2) Content-based recommenders: suggest similar items based on a particular item. This system uses item metadata, such as genre, director, description, actors, etc. for movies, to make these recommendations. The general idea behind these recommender systems is that if a person likes a particular item, he or she will also like an item that is similar to it.

Collaborative filtering engines: these systems are widely used, and they try to predict the rating or preference that a user would give an item-based on past ratings and preferences of other users.

### Demonstration of simple recommendation engine

first we download the libraries

In [11]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as pyplot
from ast import literal_eval

then we upload the movies_metadata.csv file

In [8]:
from google.colab import files
uploaded = files.upload()

Saving movies_metadata.csv to movies_metadata.csv


In [9]:
import io
metadata = pd.read_csv(io.BytesIO(uploaded['movies_metadata.csv']))
metadata.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


adult  ... vote_count
0  False  ...     5415.0
1  False  ...     2413.0
2  False  ...       92.0
3  False  ...       34.0
4  False  ...      173.0

[5 rows x 24 columns]

In [12]:
metadata['genres'] = metadata['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])


I use the IMDB Ratings to come up with our Top Movies Chart. I will use IMDB's weighted rating formula to construct my chart. Mathematically, it is represented as follows:

Weighted Rating (WR) = $(\frac{v}{v + m} . R) + (\frac{m}{v + m} . C)$

where,

v is the number of votes for the movie

m is the minimum votes required to be listed in the chart

R is the average rating of the movie

C is the mean vote across the whole report

the next is to determine value for m, the minimum votes required to be listed in the chart. We will use 95th percentile as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list.

I will build our overall Top 250 Chart and will define a function to build charts for a particular genre. Let's begin!

In [13]:
vote_counts = metadata[metadata['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = metadata[metadata['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
C

5.244896612406511

In [14]:
m = vote_counts.quantile(0.95)
m

434.0

In [15]:
metadata['year'] = pd.to_datetime(metadata['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [16]:
qualified = metadata[(metadata['vote_count'] >= m) & (metadata['vote_count'].notnull()) & (metadata['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres']]
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified.shape

(2274, 6)

Therefore, to qualify to be considered for the chart, a movie has to have at least 434 votes on IMDB. We also see that the average rating for a movie on IMDB is 5.244 on a scale of 10. 2274 Movies qualify to be on our chart.

In [17]:
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [18]:

qualified['wr'] = qualified.apply(weighted_rating, axis=1)

In [19]:
qualified = qualified.sort_values('wr', ascending=False).head(250)

In [20]:
qualified.head(15)

title  ...        wr
15480                                          Inception  ...  7.917588
12481                                    The Dark Knight  ...  7.905871
22879                                       Interstellar  ...  7.897107
2843                                          Fight Club  ...  7.881753
4863   The Lord of the Rings: The Fellowship of the Ring  ...  7.871787
292                                         Pulp Fiction  ...  7.868660
314                             The Shawshank Redemption  ...  7.864000
7000       The Lord of the Rings: The Return of the King  ...  7.861927
351                                         Forrest Gump  ...  7.860656
5814               The Lord of the Rings: The Two Towers  ...  7.851924
256                                            Star Wars  ...  7.834205
1225                                  Back to the Future  ...  7.820813
834                                        The Godfather  ...  7.814847
1154                             The Empire Strikes Back  ...  7.814099
46                                                 Se7en  ...  7.811669

[15 rows x 7 columns]

Let us now construct our function that builds charts for particular genres. For this, we will use relax our default conditions to the 85th percentile instead of 95.

In [23]:
s = metadata.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'genre'
gen_md = metadata.drop('genres', axis=1).join(s)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  """Entry point for launching an IPython kernel.


In [25]:
def build_chart(genre, percentile=0.85):
    df = gen_md[gen_md['genre'] == genre]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)
    
    qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    
    qualified['wr'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(250)
    
    return qualified

Let us see our method in action by displaying the Top 15 Romance Movies (Romance almost didn't feature at all in our Generic Top Chart despite being one of the most popular movie genres).

####Top Romance Movies

In [26]:
build_chart('Romance').head(15)

title  year  ...  popularity        wr
10309  Dilwale Dulhania Le Jayenge  1995  ...      34.457  8.565285
351                   Forrest Gump  1994  ...     48.3072  7.971357
876                        Vertigo  1958  ...     18.2082  7.811667
40251                   Your Name.  2016  ...   34.461252  7.789489
883               Some Like It Hot  1959  ...     11.8451  7.745154
1132               Cinema Paradiso  1988  ...      14.177  7.744878
19901                     Paperman  2012  ...     7.19863  7.713951
37863                  Sing Street  2016  ...   10.672862  7.689483
882                  The Apartment  1960  ...     11.9943  7.599317
38718               The Handmaiden  2016  ...   16.727405  7.566166
3189                   City Lights  1931  ...     10.8915  7.558867
24886             The Way He Looks  2014  ...     5.71127  7.331363
45437               In a Heartbeat  2017  ...    20.82178  7.003959
1639                       Titanic  1997  ...     26.8891  6.981546
19731      Silver Linings Playbook  2012  ...     14.4881  6.970581

[15 rows x 6 columns]

### Business use case

above is a simple recommender engine which lists movies based on ratings and votes included in the metadata_dataset. This is only data analysis and data wrangling.There is no machine learning algorithm involved. We can build more sophisticated recommender systems based on content using content based filtering.

We can use content based filtering and other recommender systems to recommend news articles to users based on the content of article..
In content-based filtering we use a machine learning model called cosine similarity.

We can use collaborative filtering if we can get access to personalised user data. Collaborative filtering uses powerful algorithms like Singular value decompostion.